### ISSUE TITLE
Variable output pod5 subset

### ISSUE THREAD

**Issue by Cuypers-Wim** [nanoporetech/pod5-file-format#37]
Hi,

I need to partition 1500 reads that are spread across 1375 pod5 files into 5 new pod5 files. However, each time I try, I end up with a different output. 

I made a mapping.csv file detailing read IDs and the corresponding filename I want the read to end up in. Here are the first three lines:

```bash
chikungunya_virus.pod5,2993b28e-b5f0-44dd-8612-e0fce1167e22
chikungunya_virus.pod5,27b8d110-bf05-4b78-ab4d-a5e8661343f3
chikungunya_virus.pod5,42292daf-5d66-47a5-a6d5-3900b83462dc

```

Then I run the following command

```bash
pod5 subset --threads 5 --csv mapping.csv *.pod5 
```

get this message

```bash
Subsetting 15000 read_ids into 5 outputs using 5 workers
```

and after it appears to have been completed successfully, I check the POD5 file content using:

```bash
pod5 inspect summary *.pod5
```

Issue: sometimes I end up with POD5 files with only a fraction of the reads/raws signals, while sometimes I just get an error, for example: “Failed to open pod5 file: zika_virus.pod5: IOError: Invalid signature in file”. 

When I run `pod5 subset` multiple times, I end up with different amounts of reads in each POD5 file, and always a low number of reads.

Any advice on what I should try next? 

Thanks in advance!

Wim

**Comment by HalfPhoton** [comment]
Hi @Cuypers-Wim ,

Can you please check you're on the most recent release of `pod5` with `pod5 --version`?

Is `zika_virus.pod5` a target in your `.csv` or an input from `*.pod5` ?

Kind regards,
Rich

**Comment by HalfPhoton** [comment]
Also, you reported that you want `1500` reads but the message print statement indicates `15000` reads were parsed from the `.csv` file. If this simply a typo or another issue?

Kind regards,
Rich

**Comment by Cuypers-Wim** [comment]
Hi Rich,

Thanks for looking into my question! 

I'm using `Pod5 version: 0.1.20` 

`zika_virus.pod5 ` is indeed a target in the .csv file. After `zika_virus.pod5 ` is generated by the `pod5 subset` command, inspecting it with `pod5 inspect summary` resulted in that error (sometimes).

It's indeed 15000 reads that I'm working with, 1500 was a typo :)

**Comment by HalfPhoton** [comment]
Ok, 

Would you be willing to separate the `.csv` file so that there's only 1 target and see if you get the expected result?

``` bash
# Take only zika_virus reads from the mapping
mapping.csv | grep "zika_virus" > zika_virus.csv

# See how many reads are expected in the output
wc -l zika_virus.csv
```

``` bash
mkdir output/
pod5 subset *.pod5 --output output/  --csv zika_virus.csv 

# Check the number of reads matches expectation
pod5 inspect summary output/zika_virus.csv
```

Depending on the outcome of this quick test this should inform us what the potential issue is.

Kind regards,
Rich

**Comment by HalfPhoton** [comment]
That really is `variable output` 😅 

And how many reads are you expecting? Was 213 correct?

Could you answer the following please?
- What platform are you using?
- What python version are you using?
- Pod5 version - `0.1.20`
- Is this running from a `conda` environment?
- What is the source of your input data [converted fast5s / minknow / ... ]
- Can you share the complete stdout with me?
- What is the runtime of `pod5 subset`? Is the subsetting section (showing the progress bar) very quick? 

`Invalid signature` points me towards an issue with the c_api somewhere as the files aren't being created, populated and/or closed correctly in your example. If the subsetting runtime is very short this would also indicate an issue here.

Thanks for your assistance and patience working through this.

As you have the `.csv` file, are you able to strip out the read_ids and use `pod5 filter`? Hopefully that works and will unblock your workflow?

``` bash
awk -F ',' '{print $2}' zika_virus.csv > zika_virus.txt 
pod5 filter *.pod5 --ids zika_virus.txt --output output/zika_virus.pod5 --force
```

**Comment by Cuypers-Wim** [comment]
I was expecting 3000 reads, but so far the output has never been correct. 

Here are the requested details:

- Platform: Ubuntu 18.04.5 LTS (hardware: Fujitsu server with 40 threads and 156 Gb RAM)
- Python 3.11.3
- I’m running this from a `conda` environment with `python 3.11.3` and `Pod5 0.1.20`
    - However, I had the exact same issue on a Windows desktop PC running Ubuntu on Windows. Here I did not use a conda environment.
- The source of my input data input is POD5 files generated by MinKnow version 22.12.7  (Guppy 6.4.6)
- The complete stdout (the run says it’s finished, but the generated pod5 file only contains 351 reads of the expected 3000)

```bash
Subsetting 3000 read_ids into 1 outputs using 1 workers
Finished output5/zika_virus.pod5
Done
```

stdout was empty

- `pod5 subset` ran very fast, for the subsets, the progress bar was not showing

I followed your advice to strip out the read_ids and use `pod5 filter`

`awk -F ',' '{print $2}' zika_virus.csv > zika_virus.txt 
pod5 filter *.pod5 --ids zika_virus.txt --output output/zika_virus.pod5 --force`

This didn’t work at first, resulting in the following error that is already more helpful than the previous ones: 

```bash
Subsetting please wait...

pod5 has encountered an error: 'IOError: Failed to open local file '/mnt/data1/wcuypers/pod5_200bps/FAV34252__a72e9347_e082c315_1140.pod5'. Detail: [errno 24] Too many open files'

For detailed information set POD5_DEBUG=1'
```

However, I found a temorary solution by looping over each pod5 file, using the `--missing_ok` option in `pod5 filter` , and making a new subsetted pod5 file for every original one. Then I merged them using `pod5 merge` . 

```bash

for pod5_file in $(ls *.pod5)
do
# n = unique number pod5 file
n=$(echo "$pod5_file " | sed 's/.*__.*_.*_/ /g' | sed 's/.pod5/ /g')
pod5 filter "$pod5_file" --ids zika_virus.txt --output output/zika_virus_"$n".pod5 --force --missing_ok
done

cd output
pod5 merge *.pod5 zika_merged.pod5

```

I can proceed for now with the above approach, but it is very slow :)

I hope this helps with the debugging!

**Comment by HalfPhoton** [comment]
@Cuypers-Wim 

This report has been really helpful thank you. 

It appears as though `pod5 subset` is crashing shortly after starting when it opens all the source pod5 files and the reaches the file handles `ulimit`. The output file is then either truncated or completely corrupt and the `Too many open files` error / exception is probably being obscured by multi-processing. 

This is a new presentation of the issue which has been very useful to see. Changes will be implemented in the tooling to address this issue soon. 

Out of interest how many input pod5 files are there?

The number of pod5 files / open file handles is a known issue that we're actively working to address with the minknow team. 

In the meantime, your pipeline can likely be accelerated by reverting to using `pod5 subset` and chunking over batches of a few inputs and merging those results. Increasing the `ulimit` should allow you to increase the size of the chunks too. 

Thanks again for this detailed report. We'll be working to get these issues resolved soon. 

Kind regards,
Rich

**Comment by HalfPhoton** [comment]
Hi @Cuypers-Wim ,

We've just released [0.2.0](https://github.com/nanoporetech/pod5-file-format/releases/tag/0.2.0) which reduces the number of open file handles in the tools. This will hopefully accelerate your workflow by removing the need to split and merge.

Furthermore tools now accept directory inputs which might be of use to you.

Thanks again for reporting this issue. 

Kind regards,
Rich

**Comment by HalfPhoton** [comment]
Closing as this issue should be fixed in 0.2.0
